### PR TITLE
HPCC-32171 ARM64 constant-folding support

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Added tests to regression suite, ran on Mac M2 arm64 system

### CODE REVIEW

**Comment by AttilaVamos** [comment]
On Rocky 8 there are some warnings treated as error:
```
/home/rocky/smoketest/PR-18940/HPCC-Platform/plugins/exampleplugin/exampleplugin.cpp:71:39: error: format â€˜%lluâ€™ expects argument of type â€˜long long unsigned intâ€™, but argument 6 has type â€˜uint64_tâ€™ {aka â€˜long unsigned intâ€™} [-Werror=format=]
/home/rocky/smoketest/PR-18940/HPCC-Platform/plugins/exampleplugin/exampleplugin.cpp:71:53: error: format â€˜%lldâ€™ expects argument of type â€˜long long intâ€™, but argument 10 has type â€˜int64_tâ€™ {aka â€˜long intâ€™} [-Werror=format=]
/home/rocky/smoketest/PR-18940/HPCC-Platform/plugins/exampleplugin/exampleplugin.cpp:71:58: error: format â€˜%lluâ€™ expects argument of type â€˜long long unsigned intâ€™, but argument 11 has type â€˜uint64_tâ€™ {aka â€˜long unsigned intâ€™} [-Werror=format=]
/home/rocky/smoketest/PR-18940/HPCC-Platform/plugins/exampleplugin/exampleplugin.cpp:71:63: error: format â€˜%lluâ€™ expects argument of type â€˜long long unsigned intâ€™, but argument 12 has type â€˜uint64_tâ€™ {aka â€˜long unsigned intâ€™} [-Werror=format=]
/home/rocky/smoketest/PR-18940/HPCC-Platform/plugins/exampleplugin/exampleplugin.cpp:93:39: error: format â€˜%lluâ€™ expects argument of type â€˜long long unsigned intâ€™, but argument 6 has type â€˜uint64_tâ€™ {aka â€˜long unsigned intâ€™} [-Werror=format=]
/home/rocky/smoketest/PR-18940/HPCC-Platform/plugins/exampleplugin/exampleplugin.cpp:93:53: error: format â€˜%lldâ€™ expects argument of type â€˜long long intâ€™, but argument 10 has type â€˜int64_tâ€™ {aka â€˜long intâ€™} [-Werror=format=]
/home/rocky/smoketest/PR-18940/HPCC-Platform/plugins/exampleplugin/exampleplugin.cpp:93:58: error: format â€˜%lluâ€™ expects argument of type â€˜long long unsigned intâ€™, but argument 11 has type â€˜uint64_tâ€™ {aka â€˜long unsigned intâ€™} [-Werror=format=]
/home/rocky/smoketest/PR-18940/HPCC-Platform/plugins/exampleplugin/exampleplugin.cpp:93:63: error: format â€˜%lluâ€™ expects argument of type â€˜long long unsigned intâ€™, but argument 12 has type â€˜uint64_tâ€™ {aka â€˜long unsigned intâ€™} [-Werror=format=]
cc1plus: some warnings being treated as errors
```

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: ecl/hql/hqlfold.cpp [hunk 1]
```diff
@@ -1248,7 +1248,142 @@ IValue * doFoldExternalCall(IHqlExpression* expr, unsigned foldOptions, const ch
...
+           "0: \n\t"
+               "ldr  s0,[%[vals], #0] \n\t"
+           "1: \n\t"
+               "ldr  w0,[%[sizes],#4] \n\t"
+               "cmp  w0, #4 \n\t"
+               "blt  9f \n\t"
+               "beq  0f \n\t"
+               "ldr  d1,[%[vals], #8] \n\t"
+               "b 1f \n\t"
+           "0: \n\t"
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply ecl/hql/hqlfold.cpp:1272]: They don't need to be.

File: ecl/hql/hqlfold.cpp [hunk 2]
```diff
@@ -1187,7 +1187,7 @@ IValue * doFoldExternalCall(IHqlExpression* expr, unsigned foldOptions, const ch
            );
         }
   #endif
-        assertex((len & 7) == 4);  // We need to make sure we add an ODD number of words to stack, so that it gets 8-byte aligned once pc is pushed by the call
+        assertex((len & 7) == 4);  // We need to make sure we add an ODD number of words to stack, so that it gets #8-byte aligned once pc is pushed by the call
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply ecl/hql/hqlfold.cpp:1190]: Probably (the #)

File: cmake_modules/vcpkg.cmake [hunk 3]
```diff
@@ -12,8 +12,13 @@ if(WIN32)
     set(VCPKG_HOST_TRIPLET "x64-windows" CACHE STRING "host triplet")
     set(VCPKG_TARGET_TRIPLET "x64-windows" CACHE STRING "target triplet")
 elseif(APPLE)
+  if (CMAKE_OSX_ARCHITECTURES)
+    set(VCPKG_HOST_TRIPLET "arm64-osx" CACHE STRING "host triplet")
+    set(VCPKG_TARGET_TRIPLET "arm64-osx" CACHE STRING "target triplet")
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply cmake_modules/vcpkg.cmake:17]: Makes sense

**Review by ghalliday — APPROVED** [review]
Looks good to me.  @GordonSmith ?

**Review by GordonSmith — APPROVED** [review]

**Comment by ghalliday** [comment]
@richardkchapman please squash and I will merge

**Pull request merged**